### DISCUSSION TITLE
How to use the API to manage my projects?

### DISCUSSION THREAD

**Discussion by johann-petrak** [codalab/codabench#1435] [Q&A]
I found this page: https://www.codabench.org/api/docs/ but there seems to be no documentation in the Wiki about how this can be actually used to manage my projects or how to properly authenticate. 

When I try this directly in the api docs page, all I get is "Network error"  responses: "TypeError: NetworkError when attempting to fetch resource." 

I would have hoped that perhaps using this approach would make it easier to do certain things which are not or only very tediously supported in the web interface (e.g. download all submissions is impossible for some reason, as are straightforward ways to clean up storage by deleting multiple submissions). 

When I instead try and run the curl the page shows me for an (authenticated) example request, I always get the error response: "Moved Permanently(base)"

**Comment by johann-petrak** [comment]
One problem seems to be that the documentation page shows http: instead of https: URLS and curl by default does not automatically use the https URL (wget or the browser does), so even get requests for public info fail with curl when the example command shown is used.

**Comment by johann-petrak** [comment]
However trying to use this"
```
curl -X GET "https://www.codabench.org/api/submission_scores/" -H  "accept: application/json" -H  "authorization: Basic [KEY]" -H  "X-CSRFToken: [TOKEN]"
```

where KEY and TOKEN are the values the doc page fills in after logging in, still gives this response:

`{"detail":"Authentication credentials were not provided."}`

So how do I get the credentials and then specify them correctly to see the submission scores for my projects or a specific project or all the submissions for a project/phase? 

For example I tried:
```
curl -X GET "https://www.codabench.org/api/submissions/58452/" -H  "accept: application/json" -H  "authorization: Basic [KEY]" -H  "X-CSRFToken: [TOKEN]"
```
and just got: `{"detail":"Not found."}`

**Comment by curious-broccoli** [comment]
I used the API from python for my own instance (so not sure if it works properly with codabench.org) using these as a starting point:
https://github.com/codalab/codabench/wiki/Robot-submissions
https://github.com/codalab/codabench/tree/develop/docs/example_scripts
Authenticating, as shown in the scripts, worked for me.

I mostly used them for fetching some competition data and making submissions, not managing the competition.

**Comment by johann-petrak** [comment]
Thanks @curious-broccoli the example script `get_submission_details.py` does not work for me on the codabench.org site, but it shows how to use the userid/password combination to retrieve the authentication token. Once I have that I can use it in my curl command like so, and that then works: 

```
curl -X GET "https://www.codabench.org/api/submissions/58452/" -H  "accept: application/json" -H  "authorization: Token  [TOKEN]
```

So it should be easy to then use the same request as in curl in Python ...

**Comment by johann-petrak** [comment]
To summarize the answer: 

What works for me is to first get the authentication token for my username and password like this:
```
curl -X POST "https://www.codabench.org/api/api-token-auth/" -H "Content-Type: application/json" -d '{"username":"myusername", "password":"PASSWORD"}'
```
This returns something like: `{"token":"3a116ccecea989..."}`

Then run the actual API commands similar to this, replacing `[TOKEN]` with the value that was returned for "token" in the response (without the quotes):
```
curl -X GET "https://www.codabench.org/api/submissions/58452/" -H  "accept: application/json" -H  "authorization: Token  [TOKEN]
```

Doing the equivalent in Python works as well, of course.
